### PR TITLE
ECL-552: Submit to NRS when amending returns

### PR DESCRIPTION
**Pull request by liandra-giacomine**

### CODE REVIEW

**Review by TheDanielDoyle — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/economiccrimelevyreturns/connectors/BaseConnector.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        .recoverTotal(error => Future.failed(JsResult.Exception(error)))
+  }
+
+  implicit class RequestBuilderHelpers(requestBuilder: RequestBuilder) {
+    def executeAndDeserialise[T](implicit ec: ExecutionContext, reads: Reads[T]): Future[T] =
+      requestBuilder
+        .execute[HttpResponse]
+        .flatMap { response =>
+          response.status match {
+            case OK | CREATED | ACCEPTED => response.as[T]
```
[the remark this answers is not in the document]
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyreturns/connectors/BaseConnector.scala:47]: This method deserialises the response body, NO_CONTENT wouldn't return a body right so I think for that we'd use the function below executeAndExpect

File: app/uk/gov/hmrc/economiccrimelevyreturns/controllers/ReturnSubmissionController.scala [hunk 2]
```diff
@@ -16,81 +16,99 @@
...
+      base64EncodedDmsSubmissionHtml <- checkOptionalValueExists(eclReturn.base64EncodedDmsSubmissionHtml)
+      base64EncodedNrsSubmissionHtml <- checkOptionalValueExists(eclReturn.base64EncodedNrsSubmissionHtml)
+      _                              <- nrsService
+                                          .submitToNrs(base64EncodedNrsSubmissionHtml, eclRegistrationReference, appConfig.amendReturnNotableEvent)
+                                          .asResponseError
+      eclReturnResponse              <- dmsService.submitToDms(base64EncodedDmsSubmissionHtml, Instant.now).asResponseError
+      _                               = auditService.sendReturnSubmittedEvent(eclReturn, eclRegistrationReference, None)
+    } yield eclReturnResponse
+
+  private def checkOptionalValueExists(value: Option[String]): EitherT[Future, ResponseError, String] = EitherT(
```
[the remark this answers is not in the document]
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyreturns/controllers/ReturnSubmissionController.scala:105]: Added to the base

File: app/uk/gov/hmrc/economiccrimelevyreturns/connectors/DmsConnector.scala [hunk 3]
```diff
@@ -41,24 +41,21 @@ class DmsConnector @Inject() (
...
   private def retryCondition: PartialFunction[Exception, Boolean] = {
     case e: UpstreamErrorResponse if UpstreamErrorResponse.Upstream5xxResponse.unapply(e).isDefined => true
   }
 
   def sendPdf(
     body: Source[MultipartFormData.Part[Source[ByteString, NotUsed]] with Product with Serializable, NotUsed]
-  )(implicit hc: HeaderCarrier): Future[Either[UpstreamErrorResponse, HttpResponse]] =
-    retryFor[Either[UpstreamErrorResponse, HttpResponse]]("DMS submission")(retryCondition) {
+  )(implicit hc: HeaderCarrier): Future[Unit] =
+    retryFor[Unit]("DMS submission")(retryCondition) {
```
[the remark this answers is not in the document]
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyreturns/connectors/DmsConnector.scala:54]: adding a test for it

File: app/uk/gov/hmrc/economiccrimelevyreturns/connectors/NrsConnector.scala [hunk 4]
```diff
@@ -19,41 +19,44 @@ package uk.gov.hmrc.economiccrimelevyreturns.connectors
...
-
   def submitToNrs(nrsSubmission: NrsSubmission)(implicit
     hc: HeaderCarrier
-  ): Future[NrsSubmissionResponse] =
-    retryFor[NrsSubmissionResponse]("NRS submission")(retryCondition)(
-      httpClient.POST[NrsSubmission, NrsSubmissionResponse](nrsSubmissionUrl, nrsSubmission, headers = nrsHeaders)
-    )
-
+  ): Future[NrsSubmissionResponse]                                =
+    retryFor[NrsSubmissionResponse]("NRS submission")(retryCondition) {
```
[the remark this answers is not in the document]
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyreturns/connectors/NrsConnector.scala:55]: adding a test for it

File: app/uk/gov/hmrc/economiccrimelevyreturns/services/ReturnService.scala [hunk 5]
```diff
@@ -16,55 +16,38 @@
...
-  ): Future[AuditResult] =
-    auditConnector.sendExtendedEvent(
-      ReturnSubmittedAuditEvent(
-        eclReturn,
-        eclRegistrationReference,
-        ReturnResult(Success, chargeReference, None)
-      ).extendedDataEvent
-    )(hc, ec)
+  ): EitherT[Future, ReturnsSubmissionError, SubmitEclReturnResponse] =
+    EitherT {
```
[the remark this answers is not in the document]
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyreturns/services/ReturnService.scala:36]: re-added and verifying audit calls in controller unit test for amend and first time

**Review by TheDanielDoyle — APPROVED** [review]

**Pull request merged**